Bottom: 24cf666d0f53370c43bdc86d5023950640b139ad
Top:    6205ca6c954ffe02a15ff3287b764362dd10aaea
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-12-10 00:31:01 +0000

vendor.c: Fix memory leak and unused variable in ActionLoadVendorFrom()

Also, don't bother testing x != NULL before calling free (x).


---

diff --git a/src/vendor.c b/src/vendor.c
index 39359cc..a296aa1 100644
--- a/src/vendor.c
+++ b/src/vendor.c
@@ -269,11 +269,12 @@ int
 ActionLoadVendorFrom (int argc, char **argv, int x, int y)
 {
   int i;
-  char *fname = NULL, *name = NULL;
+  char *fname = NULL;
   static char *default_file = NULL;
   char *sval;
   Resource *res, *drcres, *drlres;
   int type;
+  bool free_fname = false;
 
   cached_drill = -1;
 
@@ -291,11 +292,10 @@ ActionLoadVendorFrom (int argc, char **argv, int x, int y)
       if (fname == NULL)
 	AFAIL (load_vendor);
 
-      if (default_file != NULL)
-	{
-	  free (default_file);
-	  default_file = NULL;
-	}
+      free_fname = true;
+
+      free (default_file);
+      default_file = NULL;
 
       if (fname && *fname)
 	default_file = strdup (fname);
@@ -459,7 +459,8 @@ ActionLoadVendorFrom (int argc, char **argv, int x, int y)
 
   vendorMapEnable = true;
   apply_vendor_map ();
-  free (name);
+  if (free_fname)
+    free (fname);
   return 0;
 }
